File: src/main/java/net/lomeli/lomlib/util/ItemUtil.java
package net.lomeli.lomlib.util;

import net.minecraft.entity.item.EntityItem;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;

import net.minecraftforge.oredict.OreDictionary;

public class ItemUtil {
    public static ItemStack consumeItem(ItemStack stack) {
        if (stack.stackSize == 1) {
            if (stack.getItem().hasContainerItem(stack))
                return stack.getItem().getContainerItem(stack);
            else
                return null;
        } else {
            stack.splitStack(1);

            return stack;
        }
    }

    public static boolean itemsEqualWithMetadata(ItemStack stackA, ItemStack stackB) {
        return stackA == null ? stackB == null ? true : false : stackB != null && areItemsTheSame(stackA, stackB) && (stackA.getHasSubtypes() == false || stackA.getItemDamage() == stackB.getItemDamage());
    }

    public static boolean itemsEqualWithMetadata(ItemStack stackA, ItemStack stackB, boolean checkNBT) {
        return stackA == null ? stackB == null ? true : false : stackB != null && areItemsTheSame(stackA, stackB) && stackA.getItemDamage() == stackB.getItemDamage()
                && (!checkNBT || NBTUtil.doNBTsMatch(stackA.getTagCompound(), stackB.getTagCompound()));
    }

    public static boolean areItemsTheSame(ItemStack a, ItemStack b) {
        return a.getItem() == b.getItem() && a.getItemDamage() == b.getItemDamage();
    }

    public static ItemStack cloneStack(Item item, int stackSize) {
        if (item == null)
            return null;

        ItemStack stack = new ItemStack(item, stackSize);

        return stack;
    }

    public static ItemStack cloneStack(ItemStack stack, int stackSize) {
        if (stack == null)
            return null;

        ItemStack retStack = stack.copy();
        retStack.stackSize = stackSize;

        return retStack;
    }

    public static void dropItemStackIntoWorld(ItemStack stack, World world, double x, double y, double z, boolean velocity) {
        if (stack != null) {
            float x2 = 0.5F;
            float y2 = 0.0F;
            float z2 = 0.5F;

            if (velocity) {
                x2 = world.rand.nextFloat() * 0.8F + 0.1F;
                y2 = world.rand.nextFloat() * 0.8F + 0.1F;
                z2 = world.rand.nextFloat() * 0.8F + 0.1F;
            }
            EntityItem entity = new EntityItem(world, x + x2, y + y2, z + z2, stack.copy());

            if (velocity) {
                entity.motionX = ((float) world.rand.nextGaussian() * 0.05F);
                entity.motionY = ((float) world.rand.nextGaussian() * 0.05F + 0.2F);
                entity.motionZ = ((float) world.rand.nextGaussian() * 0.05F);
            } else {
                entity.motionY = -0.0500000007450581D;
                entity.motionX = 0.0D;
                entity.motionZ = 0.0D;
            }

            world.spawnEntityInWorld(entity);
        }
    }

    public static boolean canStacksMerge(ItemStack stack1, ItemStack stack2) {
        if (stack1 == null || stack2 == null)
            return false;
        if (!stack1.isItemEqual(stack2))
            return false;
        if (!ItemStack.areItemStackTagsEqual(stack1, stack2))
            return false;
        return true;
    }

    public static int searchForPossibleSlot(ItemStack stack, IInventory inventory) {
        for (int i = 0; i < inventory.getSizeInventory(); i++) {
            ItemStack item = inventory.getStackInSlot(i);
            if (item == null || item.getItem() == null)
                return i;
            else if (canStacksMerge(stack, item) && (item.stackSize + stack.stackSize) < item.getMaxStackSize())
                return i;
        }
        return -1;
    }

    /**
     * Attempts to put an item into the given inventory. Returns false if it fails.
     *
     * @param inventory
     * @param stack
     * @param doMove
     * @return
     */
    public static boolean insertIntoInventory(IInventory inventory, ItemStack stack, boolean doMove) {
        if (stack == null) return false;
        if (inventory == null) return false;

        int slot = searchForPossibleSlot(stack, inventory);

        if (doMove && slot > -1 && slot < inventory.getSizeInventory())
            return tryInsertStack(inventory, slot, stack, true);

        return slot != -1;
    }

    public static boolean tryInsertStack(IInventory targetInventory, int slot, ItemStack stack, boolean canMerge) {
        if (targetInventory.isItemValidForSlot(slot, stack)) {
            ItemStack targetStack = targetInventory.getStackInSlot(slot);
            if (targetStack == null) {
                int limit = targetInventory.getInventoryStackLimit();
                if (limit < stack.stackSize)
                    targetInventory.setInventorySlotContents(slot, stack.splitStack(limit));
                else {
                    targetInventory.setInventorySlotContents(slot, stack.copy());
                    stack.stackSize = 0;
                }
                return true;
            } else if (canMerge) {
                if (targetInventory.isItemValidForSlot(slot, stack) &&
                        canStacksMerge(stack, targetStack)) {
                    int space = targetStack.getMaxStackSize() - targetStack.stackSize;
                    int mergeAmount = Math.min(space, stack.stackSize);
                    ItemStack copy = targetStack.copy();
                    copy.stackSize += mergeAmount;
                    targetInventory.setInventorySlotContents(slot, copy);
                    stack.stackSize -= mergeAmount;
                    return true;
                }
            }
        }
        return false;
    }

    public static int getSlotWithStack(IInventory inventory, ItemStack stack) {
        for (int i = 0; i < inventory.getSizeInventory(); i++) {
            ItemStack item = inventory.getStackInSlot(i);
            if (item != null && OreDictionary.itemMatches(item, stack, false))
                return i;
        }
        return -1;
    }

    public static boolean hasItemStack(IInventory inventory, ItemStack stack) {
        return getSlotWithStack(inventory, stack) != -1;
    }

    public static void setEntityItemAge(EntityItem item, int age) {
        item.age = age;
    }
}

File: src/main/kotlin/net/lomeli/lomlib/util/items/ItemUtil.kt
package net.lomeli.lomlib.util.items

import net.minecraft.entity.item.EntityItem
import net.minecraft.inventory.IInventory
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.world.World

import net.minecraftforge.oredict.OreDictionary

import net.lomeli.lomlib.util.nbt.NBTUtil
import net.minecraft.init.Blocks
import net.minecraftforge.fml.common.registry.GameRegistry

object ItemUtil {
    fun consumeItem(stack: ItemStack): ItemStack? {
        if (stack.count == 1) {
            if (stack.item.hasContainerItem(stack))
                return stack.item.getContainerItem(stack)
            else
                return null
        } else {
            stack.splitStack(1)

            return stack
        }
    }

    fun itemsEqualWithMetadata(stackA: ItemStack?, stackB: ItemStack?): Boolean =
            if (stackA!!.isEmpty) stackB!!.isEmpty else (!stackB!!.isEmpty && areItemsTheSame(stackA, stackB) &&
                (stackA.hasSubtypes == false || stackA.itemDamage == stackB.itemDamage))

    fun itemsEqualWithMetadata(stackA: ItemStack?, stackB: ItemStack?, checkNBT: Boolean): Boolean {
        return if (stackA!!.isEmpty)
            stackB!!.isEmpty
        else
            !stackB!!.isEmpty && areItemsTheSame(stackA, stackB) && stackA.itemDamage == stackB.itemDamage
                    && (!checkNBT || NBTUtil.doNBTsMatch(stackA.tagCompound, stackB.tagCompound))
    }

    fun areItemsTheSame(a: ItemStack, b: ItemStack): Boolean = a.item == b.item && a.itemDamage == b.itemDamage

    fun cloneStack(item: Item?, stackSize: Int): ItemStack? {
        if (item == null && item != Item.getItemFromBlock(Blocks.AIR))
            return null

        val stack = ItemStack(item, stackSize)

        return stack
    }

    fun cloneStack(stack: ItemStack?, stackSize: Int): ItemStack? {
        if (stack!!.isEmpty)
            return null

        val retStack = stack.copy()
        retStack?.count = stackSize

        return retStack
    }

    fun dropItemStackIntoWorld(stack: ItemStack?, world: World, x: Double, y: Double, z: Double, velocity: Boolean) {
        if (!stack!!.isEmpty) {
            var x2 = 0.5f
            var y2 = 0.0f
            var z2 = 0.5f

            if (velocity) {
                x2 = world.rand.nextFloat() * 0.8f + 0.1f
                y2 = world.rand.nextFloat() * 0.8f + 0.1f
                z2 = world.rand.nextFloat() * 0.8f + 0.1f
            }
            val entity = EntityItem(world, x + x2, y + y2, z + z2, stack.copy())

            if (velocity) {
                entity.motionX = (world.rand.nextGaussian().toFloat() * 0.05f).toDouble()
                entity.motionY = (world.rand.nextGaussian().toFloat() * 0.05f + 0.2f).toDouble()
                entity.motionZ = (world.rand.nextGaussian().toFloat() * 0.05f).toDouble()
            } else {
                entity.motionY = -0.0500000007450581
                entity.motionX = 0.0
                entity.motionZ = 0.0
            }

            world.spawnEntity(entity)
        }
    }

    fun canStacksMerge(stack1: ItemStack?, stack2: ItemStack?): Boolean {
        if (stack1!!.isEmpty || stack2!!.isEmpty)
            return false
        if (!stack1.isItemEqual(stack2))
            return false
        if (!ItemStack.areItemStackTagsEqual(stack1, stack2))
            return false
        return true
    }

    fun searchForPossibleSlot(stack: ItemStack, inventory: IInventory): Int {
        for (i in 0..inventory.sizeInventory - 1) {
            val item = inventory.getStackInSlot(i)
            if (item.isEmpty)
                return i
            else if (canStacksMerge(stack, item) && item.count + stack.count < item.maxStackSize)
                return i
        }
        return -1
    }

    /**
     * Attempts to put an item into the given inventory. Returns false if it fails.

     * @param inventory
     * *
     * @param stack
     * *
     * @param doMove
     * *
     * @return
     */
    fun insertIntoInventory(inventory: IInventory?, stack: ItemStack?, doMove: Boolean): Boolean {
        if (stack!!.isEmpty) return false
        if (inventory == null) return false

        val slot = searchForPossibleSlot(stack, inventory)

        if (doMove && slot > -1 && slot < inventory.sizeInventory)
            return tryInsertStack(inventory, slot, stack, true)

        return slot != -1
    }

    fun tryInsertStack(targetInventory: IInventory, slot: Int, stack: ItemStack, canMerge: Boolean): Boolean {
        if (targetInventory.isItemValidForSlot(slot, stack)) {
            val targetStack = targetInventory.getStackInSlot(slot)
            if (targetStack.isEmpty) {
                val limit = targetInventory.inventoryStackLimit
                if (limit < stack.count)
                    targetInventory.setInventorySlotContents(slot, stack.splitStack(limit))
                else {
                    targetInventory.setInventorySlotContents(slot, stack.copy())
                    stack.count = 0
                }
                return true
            } else if (canMerge) {
                if (targetInventory.isItemValidForSlot(slot, stack) && canStacksMerge(stack, targetStack)) {
                    val space = targetStack.maxStackSize - targetStack.count
                    val mergeAmount = Math.min(space, stack.count)
                    val copy = targetStack.copy()
                    copy.count += mergeAmount
                    targetInventory.setInventorySlotContents(slot, copy)
                    stack.count -= mergeAmount
                    return true
                }
            }
        }
        return false
    }

    fun getSlotWithStack(inventory: IInventory, stack: ItemStack): Int {
        for (i in 0..inventory.sizeInventory - 1) {
            val item = inventory.getStackInSlot(i)
            if (!item.isEmpty && OreDictionary.itemMatches(item, stack, false))
                return i
        }
        return -1
    }

    fun hasItemStack(inventory: IInventory, stack: ItemStack): Boolean = getSlotWithStack(inventory, stack) != -1

    fun hasItem(inventory : IInventory, item: Item) : Boolean {
        for (i in 0..inventory.sizeInventory - 1) {
            val slotItem = inventory.getStackInSlot(i)
            if (!slotItem.isEmpty && slotItem.item == item) return true
        }
        return false
    }

    fun consumeItem(inventory : IInventory, item: Item) : Boolean {
        for (i in 0..inventory.sizeInventory - 1) {
            val slotItem = inventory.getStackInSlot(i)
            if (!slotItem.isEmpty && slotItem.item == item) {
                slotItem.count--
                if (slotItem.count <= 0)
                    inventory.setInventorySlotContents(i, ItemStack.EMPTY)
                return true
            }
        }
        return false
    }

    fun setEntityItemAge(item: EntityItem, age: Int) {
        item.age = age
    }

    fun registerItem(item : Item, name : String) {
        item.setRegistryName(name)
        GameRegistry.register(item)
    }
}
